src/server/game/Combat/NewThreatManager.cpp

```diff
@@ -0,0 +1,394 @@
+    {
+        if (!ref->IsOnline())
+            continue;
+        data << ref->GetVictim()->GetPackGUID();
+        data << uint32(ref->GetThreat() * 100);
+        ++count;
+    }
+    data.put<uint32>(countPos, count);
+    _owner->SendMessageToSet(&data, false);
+}
+
+void NewThreatManager::ScaleMyThreat(float factor)
+{
+    factor = std::max<float>(factor, 0.0f);
+    for (auto const& pair : _threatenedByMe)
+        pair.second->ScaleThreat(factor);
+}
+
+void NewThreatManager::RemoveMeFromThreatLists()
+{
+    while (!_threatenedByMe.empty())
+        _threatenedByMe.begin()->second->ClearThreat();
+}
+
+void NewThreatManager::UpdateMyTempModifiers()
+{
+    int32 mod = 0;
+    for (AuraEffect const* eff : _owner->GetAuraEffectsByType(SPELL_AURA_MOD_TOTAL_THREAT))
+        mod += eff->GetAmount();
+
+    for (auto const& pair : _threatenedByMe)
+        pair.second->_tempModifier = mod;
+}
+
+void NewThreatManager::ForwardThreatForAssistingMe(Unit* assistant, float baseAmount, SpellInfo const* spell, bool ignoreModifiers)
+{
+    for (auto const& pair : _threatenedByMe)
+        pair.second->GetOwner()->GetThreatManager().AddThreat(assistant, baseAmount, spell, ignoreModifiers);
+}
+
+void NewThreatManager::sendRemoveToClients(Unit const* victim) const
+{
+    WorldPacket data(SMSG_THREAT_REMOVE, 16);
+    data << _owner->GetPackGUID();
+    data << victim->GetPackGUID();
+    _owner->SendMessageToSet(&data, false);
+}
+
+void NewThreatManager::sendNewVictimToClients(ThreatReference const* victimRef) const
+{
+    WorldPacket data(SMSG_HIGHEST_THREAT_UPDATE, (_sortedThreatList.size() + 2) * 8);
+    data << _owner->GetPackGUID();
+    data << victimRef->_victim->GetPackGUID();
+    size_t countPos = data.wpos();
+    data << uint32(0); // placeholder
+    uint32 count = 0;
+    for (ThreatReference* ref : _sortedThreatList)
+    {
+        if (!ref->IsOnline())
+            continue;
+        data << ref->GetVictim()->GetPackGUID();
+        data << uint32(ref->GetThreat() * 100);
+        ++count;
+    }
+    data.put<uint32>(countPos, count);
+    _owner->SendMessageToSet(&data, false);
+}
+
+void NewThreatManager::putThreatListRef(ObjectGuid const& guid, ThreatReference* ref)
+{
+    auto& inMap = _myThreatListEntries[guid];
+    ASSERT(!inMap && "Duplicate threat list entry being inserted - memory leak!");
+    inMap = ref;
+    _sortedThreatList.push_back(ref);
+}
+
+void NewThreatManager::purgeThreatListRef(ObjectGuid const& guid, bool sendRemove)
+{
+    auto it = _myThreatListEntries.find(guid);
+    if (it == _myThreatListEntries.end())
+        return;
+    ThreatReference* ref = it->second;
+    _myThreatListEntries.erase(it);
+
+    if (_currentVictimRef == ref)
+        _currentVictimRef = nullptr;
+    
+    // start at back since that's what ClearThreat does (this makes ClearThreat take linear instead of quadratic time)
+    for (size_t i = _sortedThreatList.size(); (i--) > 0;)
```

Comment:
It's not sorted until it needs to be (which isn't now). Swap and pop is reasonable though.